start: Delay pull secret on disk check to end

praveenkumar opened the pull request:
In current scenario MCO is patched with user provided pull secret and just after, we check if pull secret is part of disk which takes ~1 min since MCD make that change and the report to MCO. In this PR we are delaying this pull secret check to end because instead of blocking it for ~1 min better to execute other part and at the end check if pull secret is part of disk image.

With this PR (crc start time, 6 runs)
```
real	4m9.247s
user	0m0.557s
sys	0m0.165s

real	4m0.455s
user	0m0.619s
sys	0m0.168s

real	4m5.962s
user	0m0.445s
sys	0m0.154s

real	3m59.594s
user	0m0.661s
sys	0m0.179s

real	4m3.958s
user	0m0.563s
sys	0m0.177s

real	4m28.806s
user	0m0.460s
sys	0m0.171s
```

Without this PR
```
real	5m7.235s
user	0m0.797s
sys	0m0.181s

real	4m28.741s
user	0m0.891s
sys	0m0.195s

real	6m6.815s
user	0m0.747s
sys	0m0.194s

real	5m1.733s
user	0m0.395s
sys	0m0.199s

real	4m30.551s
user	0m0.466s
sys	0m0.173s

real	4m31.067s
user	0m0.673s
sys	0m0.183s
```

## Description

<!--
Thank you for your pull request (PR)!

Please provide a description of what your PR does providing a link (if applicable) to the issue it fixes.
-->

Fixes: #N

Relates to: #N, PR #N, ...

<!--
Describe in plain English what you solved and how. For instance, _Added `start` command to CRC so the user can create a VM and set up a single-node OpenShift cluster on it with one command. It requires blablabla..._
-->

## Type of change
<!--
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change
- [ ] Chore (non-breaking change which doesn't affect codebase;
  test, version modification, documentation, etc.)

## Proposed changes
<!--
List main as well as consequential changes you introduced or had to introduce.

1. Add `start` command.
2. Add `setup` as prerequisite to `start`.
3. ...
-->

## Testing
<!--
What is the _bottom-line_ functionality that needs testing? Describe in pseudo-code or in English. Use verifiable statements that tie your changes to existing functionality.

1. `start` succeeds first time after `setup` succeeded
2. stdout contains ... if `start` succeeded
3. stderr contains ... after `start` failed
4. `status` returns ... if `start` succeeded
5. `status` returns ... if `start` failed
6. `start` fails after `start` succeeded or after `status` says CRC is `Running`
7. ...
-->

## Contribution Checklist
- [x] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
- [x] I have performed a self-review of my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Which platform have you tested the code changes on? <!-- Only put an `x` in applicable platforms -->
    - [x] Linux
    - [ ] Windows
    - [ ] MacOS

## Summary by Sourcery

Enhancements:
- Move pull secret disk presence verification to the end of the start process to prevent early blocking and improve performance

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Bug Fixes**
  - Improved reliability when accessing the pull secret by ensuring connectivity checks are performed before each attempt.
  - Adjusted the sequence of operations during cluster startup to wait for cluster stabilization before verifying the presence of the pull secret.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

praveenkumar committed:
start: Delay pull secret on disk check to end

In current scenario MCO is patched with user provided pull secret and
just after, we check if pull secret is part of disk which takes ~1 min
since MCD make that change and the report to MCO. In this PR we are
delaying this pull secret check to end because instead of blocking it
for ~1 min better to execute other part and at the end check if pull
secret is part of disk image.

With this PR (crc start time, 6 runs)
```
real	4m9.247s
user	0m0.557s
sys	0m0.165s

real	4m0.455s
user	0m0.619s
sys	0m0.168s

real	4m5.962s
user	0m0.445s
sys	0m0.154s

real	3m59.594s
user	0m0.661s
sys	0m0.179s

real	4m3.958s
user	0m0.563s
sys	0m0.177s

real	4m28.806s
user	0m0.460s
sys	0m0.171s
```

Without this PR
```
real	5m7.235s
user	0m0.797s
sys	0m0.181s

real	4m28.741s
user	0m0.891s
sys	0m0.195s

real	6m6.815s
user	0m0.747s
sys	0m0.194s

real	5m1.733s
user	0m0.395s
sys	0m0.199s

real	4m30.551s
user	0m0.466s
sys	0m0.173s

real	4m31.067s
user	0m0.673s
sys	0m0.183s
```

praveenkumar commented:
/retest

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -614,6 +610,10 @@ func (client *client) Start(ctx context.Context, startConfig types.StartConfig)
 		logging.Warnf("Cluster is not ready: %v", err)
 	}
 
+	if err := cluster.WaitForPullSecretPresentOnInstanceDisk(ctx, sshRunner); err != nil {
```
can `cluster.StartMonitoring()` and `cluster.WaitForClusterStable` succeed if the pull secret is not written yet to disk?

praveenkumar replied:
We don't have monitoring related images cached in the bundle so when user want to use monitoring operator then they have to wait for images and images can will be downloaded using the pull secret but in our case we don't dictate when this is available on the disk since MCO is responsible for it. Now assume when monitoring operator start and pull secret is not yet available on disk then it will fail and again retry but eventually succeed as soon as images are pulled.

praveenkumar replied:
@cfergeau any thought on that?

cfergeau replied:
If there is some code which needs the pull secret to be on disk to work, and we have a "wait until it works" function for this code, then I’d prefer if `cluster.WaitForPullSecretPresentOnInstanceDisk` was called before it.
Otherwise the "wait until it works" function will also act as a hidden `cluster.WaitForPullSecretPresentOnInstanceDisk`, and `cluster.WaitForPullSecretPresentOnInstanceDisk` will almost be a no-op.

praveenkumar replied:
@cfergeau We don't have any code that directly depends on the pull secret being present on disk. Instead, we're executing a set of `oc` commands, some action (like monitoring or marketplace) involve pulling container images. These operations require the pull secret, and if it's not available, they can result in image pull errors. However, these errors are non-blocking and are retried, so they don't cause a complete failure.

Our current goal in this PR is to optimize startup time by checking for the presence of the pull secret on disk only at the end of the process. In fact, we could go a step further and consider removing this check entirely. Once the pull secret is patched, it's the responsibility of the Machine Config Operator (MCO) to propagate it to the disk. Our check is effectively a redundant safeguard as of now.

praveenkumar committed:
cluster: wait for ssh connectivity before check for pull-secret

During CI failure it is observe sometime following error happen when
checking the pull secret present on the disk. This PR make sure that ssh
connectivity is present before checking pull-secret.
```
DEBU SSH command results: err: ssh: unexpected packet in response to channel open: <nil>, output:
DEBU error: Temporary error: ssh command error:
```

praveenkumar commented:
/retest

praveenkumar commented:
@coderabbitai review

praveenkumar commented:
@coderabbitai full review

cfergeau commented:
This probably can move forward?

praveenkumar commented:
> This probably can move forward?

It need the approval, from my end it is good to go.
